### DISCUSSION TITLE
Can't get the same AC as with BP-tools

### DISCUSSION THREAD

**Discussion by mjgalindo** [knovichikhin/pyemv#7] [Q&A] [answered]
I'm trying to pyemv to automate some tasks that I'd usually do with bp-tools. I get the same session keys with pyemv as I do with bp-tools, but when I generate the AC, I can't get them to match, even if I use a simple session key. Is there a reason for this?

**Comment by knovichikhin** [comment, marked as answer]
There could be several reasons.

1. The application does not use AC session key for AC generation. It's possible that this particular application uses ICC AC key. 
2. The input into the AC function is not correct or not in correct order.
a. You need to provide exact list of tags in exact order concatenated together into the generate AC function.
b. Some applications may need portions of some tags (e.g. CVR  or CVR and counters  out of 9F10). 
3. Chosen padding method is not correct. There are two padding methods.
a. One with a bunch of 0x00 added at the end.
b. One with 0x80 control bit and then a bunch of 0x00 added at the end. Most modern applications use this one.

If you have documentation on how your application AC is calculated then it'd be easy to find where the problem is. 
If your application is one of the well-known ones, you can take a look at `pyemv/cvn.py` and `pyemv/tests/test_cvn_hsm.py` for a sample of how to produce an AC. `cvn.py` module pulls everything together and shows how things are connected.

If you can share more info maybe I can help more. These are some general pointers.

**Reply by mjgalindo** [reply]
Aha! The cvn module does produce the same results. I was indeed using CVN10 from BP-tools. I misunderstood the use of generate_ac thinking that was the tool for the end-user. Thanks!
